File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder)
    {
        return treeBuilder(preorder, inorder, 0, (int)inorder.size() - 1, 0);
    }

    TreeNode *treeBuilder(vector<int> &preorder, vector<int> &inorder, int inStart = 0, int inEnd = -1, int preIt = 0)
    {
        if (preIt > preorder.size() - 1 || inStart > inEnd)
            return nullptr;

        TreeNode *root = new TreeNode(preorder[preIt]);
        int idx=0;
        for (int i = inStart; i <= inEnd; i++)
        {
            if (inorder[i] == root->val)
            {
            idx=i;
            }

        }
        
root->left = treeBuilder(preorder, inorder, inStart, idx - 1, preIt + 1);
root->right = treeBuilder(preorder, inorder, idx + 1, inEnd, preIt + (idx - inStart + 1));

        return root;
    }
};

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.java
class Solution {
    private int preIdx;
    private Map<Integer, Integer> inorderIndexMap;

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        preIdx = 0; // start from the first element in preorder

        // Map inorder values -> indices for quick lookup
        inorderIndexMap = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) {
            inorderIndexMap.put(inorder[i], i);
        }

        return build(preorder, 0, inorder.length - 1);
    }

    private TreeNode build(int[] preorder, int inLeft, int inRight) {
        if (inLeft > inRight)
            return null;

        // root from preorder
        int rootVal = preorder[preIdx++];
        TreeNode root = new TreeNode(rootVal);

        // find root in inorder
        int idx = inorderIndexMap.get(rootVal);

        // left then right (because preorder is root-left-right)
        root.left = build(preorder, inLeft, idx - 1);
        root.right = build(preorder, idx + 1, inRight);

        return root;
    }
}
